Added support for live reload for any template, static files.

Cees-Kettenis opened the pull request:
1. Added dev_watcher.go, it implements a file system watcher that monitors changes and assigns hashes to files for efficient change detection.

2. Added reload.go,  handles subscription and notification logic for broadcasting file change events to connected clients.

3. Updated server.go, Integrates the new live reload functionality, enabling it only when dev = true in config.toml.

4. Modified routes.go, adjusted static file handling to optionally bypass caching, allowing files to be reloaded during development.

5.Updated footer.gohtml, added JavaScript logic to establish a connection to the new reload endpoint for receiving live update messages.

CeesKettenis committed:
1. Added dev_watcher.go, it implements a file system watcher that monitors changes and assigns hashes to files for efficient change detection.

2. Added reload.go,  handles subscription and notification logic for broadcasting file change events to connected clients.

3. Updated server.go, Integrates the new live reload functionality, enabling it only when dev = true in config.toml.

4. Modified routes.go, adjusted static file handling to optionally bypass caching, allowing files to be reloaded during development.

5.Updated footer.gohtml, added JavaScript logic to establish a connection to the new reload endpoint for receiving live update messages.

topi314 commented on server/web/templates/footer.gohtml:
```diff
@@ -16,6 +16,31 @@
+                connect();
+            })();
+        </script>
+    {{ end }}
```
ideally this would be it's own template in `templates.gohtml`

topi314 commented on server/reload.go:
```diff
@@ -0,0 +1,93 @@
+	if notifier.closed {
+		ch := make(chan struct{})
+		close(ch)
+		return -1, ch
```
you can just return a `nil` channel here

topi314 commented on server/server.go:
```diff
@@ -82,6 +91,16 @@ func New(cfg Config) (*Server, error) {
 
 	httpClient := &http.Client{}
+	var (
+		reloadNotifier *reloadNotifier
```
either export the `reloadNotifier` type or name the variable different since you are shadowing the type with this variable

CeesKettenis committed:
1. named handeler, response and writer by 1 letter variables.
2. no longer copy all the funcs into a temp func i just call Funcs(devTemplateFuncs)
3. Reworked Subscribe() to take in a func so we no longer need unsubscribe.
4. renamed reloadNotifier "mutex" to "mu"

Cees-Kettenis commented on server/reload.go:
```diff
@@ -0,0 +1,93 @@
+// Subscribe registers a new listener and returns both its ID and a channel that
+// delivers reload signals. If the notifier has already been closed we return a
+// closed channel so callers can fail fast.
+func (notifier *reloadNotifier) Subscribe() (int, <-chan struct{}) {
```
Reworked this function as you suggested.

CeesKettenis committed:
1. changed reloadNotifier to devLiveReloadNotifier to avoid confusion.
2.  moved the javascript to templates.gohtml under "live_reload_script"
3. updated the footer to use this new template.

topi314 reviewed (approved):
good stuff
